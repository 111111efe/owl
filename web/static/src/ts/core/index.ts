export { QWeb } from "./qweb_vdom";
export { EventBus } from "./event_bus";
export { Component } from "./component";
