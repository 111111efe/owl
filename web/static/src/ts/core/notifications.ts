import { EventBus as Bus } from "../core/event_bus";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

export interface INotification {
  id: number;
  title: string;
  message: string;
  type: "notification" | "warning";
  sticky: boolean;
}

export type NotificationEvent = "notifications_updated";

export type Callback = (notifs: INotification[]) => void;

export interface INotificationManager {
  add(notif: Partial<INotification>): number;
  close(id: number): void;
  on(event: NotificationEvent, owner: any, callback: Callback): void;
}

//------------------------------------------------------------------------------
// Notification Manager
//------------------------------------------------------------------------------

export class NotificationManager extends Bus implements INotificationManager {
  nextID = 1;
  notifications: INotification[] = [];

  add(notif: Partial<INotification>): number {
    const id = this.nextID++;
    const defaultVals = {
      title: "",
      message: "",
      type: "notification",
      sticky: false
    };
    const notification = Object.assign(defaultVals, notif, { id });
    this.notifications.push(notification);
    this.trigger("notifications_updated", this.notifications);
    if (!notification.sticky) {
      setTimeout(() => this.close(id), 2500);
    }
    return id;
  }
  close(id: number) {
    this.notifications = this.notifications.filter(n => n.id !== id);
    this.trigger("notifications_updated", this.notifications);
  }
}
