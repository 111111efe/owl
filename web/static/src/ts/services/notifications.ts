import { EventBus as Bus } from "../core/event_bus";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

export interface INotification {
  id: number;
  title: string;
  message: string;
  type: "notification" | "warning";
  sticky: boolean;
}

export type NotificationEvent = "notification_added" | "notification_removed";

export type Callback = (notif: INotification) => void;

export interface INotificationManager {
  add(notif: Partial<INotification>): number;
  close(id: number): void;
  on(event: NotificationEvent, owner: any, callback: Callback): void;
}

//------------------------------------------------------------------------------
// Notification Manager
//------------------------------------------------------------------------------
export class NotificationManager extends Bus implements INotificationManager {
  nextID = 1;
  notifications: { [key: number]: INotification } = {};

  add(notif: Partial<INotification>): number {
    const id = this.nextID++;
    const defaultVals = {
      title: "",
      message: "",
      type: "notification",
      sticky: false
    };
    const notification = Object.assign(defaultVals, notif, { id });
    this.notifications[id] = notification;
    this.trigger("notification_added", notification);
    if (!notification.sticky) {
      setTimeout(() => this.close(id), 2500);
    }
    return id;
  }
  close(id: number) {
    let notification = this.notifications[id];
    if (notification) {
      delete this.notifications[id];
      this.trigger("notification_removed", notification);
    }
  }
}
