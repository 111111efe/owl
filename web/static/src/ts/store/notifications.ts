import { EventBus as Bus } from "../core/event_bus";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

export interface INotification {
  id: number;
  title: string;
  message: string;
  type: "notification" | "warning";
  sticky: boolean;
}

export interface INotificationManager {
  add(notif: Partial<INotification>): number;
  close(id: number): void;
  on(
    event: "notification_added",
    owner: any,
    callback: (notif: INotification) => void
  ): void;
  on(
    event: "notification_closed",
    owner: any,
    callback: (id: number) => void
  ): void;
}

//------------------------------------------------------------------------------
// Notification Manager
//------------------------------------------------------------------------------

export class NotificationManager extends Bus implements INotificationManager {
  nextID = 1;

  add(notif: Partial<INotification>): number {
    const id = this.nextID++;
    const defaultVals = {
      title: "",
      message: "",
      type: "notification",
      sticky: false
    };
    const notification = Object.assign(defaultVals, notif, { id });
    this.trigger("notification_added", notification);
    if (!notification.sticky) {
      setTimeout(() => this.close(id), 2500);
    }
    return id;
  }
  close(id: number) {
    this.trigger("notification_closed", id);
  }
}
