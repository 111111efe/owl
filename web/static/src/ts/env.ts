import { WEnv } from "./core/component";
import { QWeb } from "./core/qweb_vdom";
import { Registry } from "./core/registry";
import { idGenerator, memoize } from "./core/utils";
import { TemplateLoader } from "./loaders/templates";
import { actionRegistry } from "./registries";
import { RPC } from "./services/ajax";
import {
  ActionManager,
  ActionWidget,
  IActionManager
} from "./store/action_manager";
import { Ajax, IAjax } from "./store/ajax";
import {
  INotificationManager,
  NotificationManager
} from "./store/notifications";
import { IRouter, Router } from "./store/router";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

export interface Env extends WEnv {
  // services
  actionManager: IActionManager;
  ajax: IAjax;
  notifications: INotificationManager;
  router: IRouter;

  // registries
  actionRegistry: Registry<ActionWidget>;

  // helpers
  rpc: IAjax["rpc"];

  // configuration
  debug: boolean;
  isMobile: boolean;
}

interface Loaders {
  loadTemplates: TemplateLoader;
}

interface Services {
  rpc: RPC;
}

type EnvBuilder = (loaders: Loaders, services: Services) => Promise<Env>;
//------------------------------------------------------------------------------
// Environment
//------------------------------------------------------------------------------

/**
 * init returns the main environment for the application.
 *
 * Note that it does not make much sense (except for tests) to have more than
 * one environment. For example, with two environment, the router code in one
 * environment will probably interfere with the code from the other environment.
 *
 * For this reason, the result of init is memoized: every call to
 * this function will actually return the same environment.
 */
export const makeEnv: EnvBuilder = memoize(_makeEnv);

async function _makeEnv(loaders: Loaders, services: Services): Promise<Env> {
  // services
  const qweb = new QWeb();
  const router = new Router();
  const ajax = new Ajax(services.rpc);
  const actionManager = new ActionManager(actionRegistry, ajax);
  const notifications = new NotificationManager();

  // templates
  const templates = await loaders.loadTemplates();
  qweb.addTemplate("default", "<div/>");
  qweb.loadTemplates(templates);

  const env: Env = {
    // Base widget requirements
    qweb,
    getID: idGenerator(),

    actionManager,
    ajax,
    notifications,
    actionRegistry,
    router,

    rpc: ajax.rpc,

    debug: false,
    isMobile: window.innerWidth <= 768
  };
  return env;
}
