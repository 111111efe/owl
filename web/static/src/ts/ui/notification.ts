import { INotification } from "../store/store";
import { Widget } from "./widget";

export class Notification extends Widget<INotification, {}> {
  template = "web.notification";

  close(ev: MouseEvent) {
    // we do not want the url to change
    ev.preventDefault();
    this.env.closeNotification(this.props.id);
  }
}
