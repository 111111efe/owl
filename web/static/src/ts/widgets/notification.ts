import { INotification } from "../store";
import { Widget } from "./widget";

export class Notification extends Widget<INotification, {}> {
  template = "web.notification";

  close(ev: MouseEvent) {
    // we do not want the url to change
    ev.preventDefault();
    this.env.dispatch("close_notification", this.props.id);
  }
}
