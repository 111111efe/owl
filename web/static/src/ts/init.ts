import { Ajax } from "./store/ajax";
import { NotificationManager } from "./core/notifications";
import { QWeb } from "./core/qweb_vdom";
import { Router } from "./core/router";
import { findInTree, idGenerator, memoize } from "./core/utils";
import { actionRegistry } from "./registries";
import { ActionManager } from "./store/action_manager";
import { MenuInfo, MenuItem } from "./widgets/root";
import { Env } from "./widgets/widget";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

interface InitializedData {
  env: Env;
  menuInfo: MenuInfo;
}

/**
 * init returns the main environment for the application.
 *
 * Note that it does not make much sense (except for tests) to have more than
 * one environment. For example, with two environment, the router code in one
 * environment will probably interfere with the code from the other environment.
 *
 * For this reason, the result of init is memoized: every call to
 * this function will actually return the same environment.
 */
export const init = memoize(async function(): Promise<InitializedData> {
  // services
  const qweb = new QWeb();
  const router = new Router();
  const ajax = new Ajax(ajaxFetch);
  const actionManager = new ActionManager(actionRegistry, ajax);
  const notifications = new NotificationManager();

  // templates
  const templates = await loadTemplates();
  qweb.addTemplate("default", "<div/>");
  qweb.loadTemplates(templates);

  const menuInfo = loadMenus();

  const env: Env = {
    // Base widget requirements
    qweb,
    getID: idGenerator(),

    actionManager,
    ajax,
    notifications,
    actionRegistry,
    router,

    rpc: ajax.rpc,

    debug: false,
    isMobile: window.innerWidth <= 768
  };
  return { env, menuInfo };
});

//------------------------------------------------------------------------------
// Adapters
//------------------------------------------------------------------------------

function ajaxFetch(route: string, params: any): Promise<any> {
  console.log("RPC", route, params);
  const delay = Math.random() * 150;
  return new Promise(resolve => setTimeout(resolve, delay));
}

/**
 * Load xml templates as a string.
 */
async function loadTemplates(): Promise<string> {
  const result = await fetch("templates.xml");
  if (!result.ok) {
    throw new Error("Error while fetching xml templates");
  }
  return result.text();
}

/**
 * Load all menu items
 */
function loadMenus(): MenuInfo {
  const menuItems: BaseMenuItem[] = (<any>window).odoo.menus;
  if (!menuItems) {
    throw new Error("Cannot find menus description");
  }
  const menuInfo = getMenuInfo(menuItems);
  delete (<any>window).odoo.menus; // overkill?
  return menuInfo;
}

interface BaseMenuItem {
  id: number;
  name: string;
  parent_id: number | false;
  action: string | false;
  icon: string | false;
  children: BaseMenuItem[];
}

/**
 * Generate a valid MenuInfo object from a list of BaseMenuItems. This function
 * is supposed to be called once at startup.
 */
export function getMenuInfo(items: BaseMenuItem[]): MenuInfo {
  const menus: { [key: number]: MenuItem | undefined } = {};
  const actionMap: { [id: number]: MenuItem | undefined } = {};
  const roots: number[] = [];

  // build MenuItems
  for (let root of items) {
    roots.push(root.id);
    addToMap(root);
  }

  function addToMap(m: BaseMenuItem, app?: MenuItem): MenuItem {
    const item: Partial<MenuItem> = {
      id: m.id,
      name: m.name,
      parentId: m.parent_id,
      action: m.action,
      icon: m.icon,
      actionId: -1 // will be filled later on with correct id
    };
    item.app = app || (item as MenuItem);
    item.children = m.children.map(c => addToMap(c, item.app as MenuItem));
    menus[item.id!] = item as MenuItem;
    return item as MenuItem;
  }

  // add proper actionId to every menuitems
  for (let menuId in menus) {
    const menu = menus[menuId]!;
    let menuWithAction = menu && findInTree(menu, m => Boolean(m.action));
    if (menuWithAction) {
      menu.actionId = parseInt(
        (<string>menuWithAction.action).split(",")[1],
        10
      );
      if (menuWithAction === menu) {
        actionMap[menu.actionId] = menu;
      }
    }
  }

  return { menus, roots, actionMap };
}
