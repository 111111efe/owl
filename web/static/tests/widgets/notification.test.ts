import { Env } from "../../src/ts/env";
import { INotification } from "../../src/ts/services/notifications";
import { Notification } from "../../src/ts/widgets/notification";
import { makeTestEnv, makeTestFixture } from "../helpers";

//------------------------------------------------------------------------------
// Setup and helpers
//------------------------------------------------------------------------------

let fixture: HTMLElement;
let env: Env;

beforeEach(() => {
  fixture = makeTestFixture();
  env = makeTestEnv();
});

afterEach(() => {
  fixture.remove();
});

function makeNotification(notif: Partial<INotification> = {}): INotification {
  const defaultNotif = {
    id: 1,
    title: "title",
    message: "message",
    type: "notification",
    sticky: false
  };
  return Object.assign(defaultNotif, notif);
}

//------------------------------------------------------------------------------
// Tests
//------------------------------------------------------------------------------

test("can be rendered", async () => {
  const notif = makeNotification({ title: "title", message: "message" });
  const navbar = new Notification(env, notif);
  await navbar.mount(fixture);
  expect(fixture.innerHTML).toMatchSnapshot();
});

test("can be closed by clicking on it (if sticky)", async () => {
  let notif: INotification;
  let removed = false;
  env.notifications.on("notification_added", null, _notif => (notif = _notif));
  env.notifications.add({
    title: "title",
    message: "message",
    sticky: true
  });
  env.notifications.on("notification_removed", null, () => (removed = true));

  const navbar = new Notification(env, notif!);
  await navbar.mount(fixture);
  expect(removed).toBe(false);
  (<any>fixture.getElementsByClassName("o_close")[0]).click();
  expect(removed).toBe(true);
});
