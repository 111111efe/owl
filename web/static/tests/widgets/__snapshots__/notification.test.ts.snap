// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can be rendered 1`] = `
"<div class=\\"o_notification\\">
        
        <div class=\\"o_notification_title\\">title</div>
        <div class=\\"o_notification_content\\">message</div>

    </div>"
`;
